File: numbers.py
import math


# Gauss–Legendre algorithm, limit = 100 digits
def find_pi_to_nth_digit(n):
    if n > 100:
        raise ValueError("The number of decimal places is limited to 100.")

    accuracy = 10 ** (-n)

    a = 1.0
    b = 1.0 / (2 ** 0.5)
    t = 0.25
    p = 1.0

    while abs(a - b) > accuracy:
        a_next = (a + b) / 2
        b = (a * b) ** 0.5
        t -= p * ((a - a_next) ** 2)
        a = a_next
        p *= 2

    pi = ((a + b) ** 2) / (4 * t)
    return round(pi, n)


# Taylor series, limit = 100 digits
def find_e_to_nth_digit(n):
    if n > 100:
        raise ValueError("The number of decimal places is limited to 100.")

    accuracy = 10 ** (-n)

    e = 0.0
    factorial = 1.0
    k = 0

    while True:
        e += (1.0 / factorial)
        if (1.0 / factorial) < accuracy:
            break
        k += 1
        factorial *= k

    return round(e, n)


# Fibonacci Sequence to n
def fibonacci_sequence(n):
    if n < 0:
        raise ValueError("The number must be a non-negative integer.")
    fib = [0, 1]
    i = 2
    while True:
        next = fib[i - 1] + fib[i - 2]
        if next > n:
            break
        fib.append(next)
        i += 1
    return fib


# Fibonacci Sequence to nth number
def fibonacci_nth_number(n):
    if n < 0:
        raise ValueError("The number must be a non-negative integer.")
    if n == 0:
        return []
    elif n == 1:
        return [0]

    fib = [0, 1]
    for i in range(2, n):
        fib.append(fib[i - 1] + fib[i - 2])

    return fib


# Find all prime factors
def prime_factors(n):
    if n < 2:
        raise ValueError("The number must be a positive integer greater than 1.")
    factors = []
    while n % 2 == 0:
        factors.append(2)
        n //= 2
    limit = int(n ** 0.5) + 1
    for i in range(3, limit, 2):
        while n % i == 0:
            factors.append(i)
            n //= i

    if n > 2:
        factors.append(n)

    return factors


# Find Cost of Tile to Cover W x H Floor
def tile_cost(w, h, c):
    if w < 0 or h < 0 or c < 0:
        raise ValueError("All dimensions must be non-negative.")
    return w * h * c


# Calculate monthly payments of a fixed term mortgage over given Nth terms at a given interest rate
def mortgage_calculator_monthly_payments(principal, annual_rate, years, interval):
    if principal < 0 or annual_rate < 0 or years < 0:
        raise ValueError("All parameters must be non-negative.")
    if interval == "monthly":
        n = years * 12
        r = annual_rate / 12
    elif interval == "weekly":
        n = years * 52
        r = annual_rate / 52
    elif interval == "daily":
        n = years * 365
        r = annual_rate / 365
    else:
        raise ValueError("Invalid compounding interval. Choose 'monthly', 'weekly' or 'daily'.")

    if r == 0:
        monthly_payment = principal / n
    else:
        monthly_payment = principal * (r * (1 + r) ** n) / (((1 + r) ** n) - 1)

    return round(monthly_payment, 2)


# Calculate how long it will take the user to pay back the loan
def mortgage_calculator_payback_time(principal, annual_rate, monthly_payment, interval):
    if principal < 0 or annual_rate < 0 or monthly_payment < 0:
        raise ValueError("All parameters must be non-negative.")
    if interval == "monthly":
        r = annual_rate / 12
    elif interval == "weekly":
        r = annual_rate / 52
    elif interval == "daily":
        r = annual_rate / 365
    else:
        raise ValueError("Invalid compounding interval. Choose 'monthly', 'weekly' or 'daily'.")

    if r == 0:
        payback_time = principal / monthly_payment / 12
    else:
        payback_time = math.log(monthly_payment / (monthly_payment - r * principal)) / math.log(1 + r)

    return payback_time


# Figure out the change and the number of banknotes, quarters, dimes, nickels, pennies needed for the change
def change_return(n):
    if n < 0:
        raise ValueError("The amount must be a non-negative integer.")
    change = {
        "hundred-dollar bills": 0, "fifty-dollar bills": 0, "twenty-dollar bills": 0, "ten-dollar bills": 0,
        "five-dollar bills": 0, "two-dollar bills": 0, "one-dollar bills": 0,
        "quarters": 0, "dimes": 0, "nickels": 0, "pennies": 0
    }
    values = {
        100: "hundred-dollar bills", 50: "fifty-dollar bills", 20: "twenty-dollar bills", 10: "ten-dollar bills",
        5: "five-dollar bills", 2: "two-dollar bills", 1: "one-dollar bills", 0.25: "quarters",
        0.1: "dimes", 0.05: "nickels", 0.01: "pennies"
    }

    for value, name in values.items():
        if n >= value:
            change[name] = int(n / value)
            n = n % value

    return [[key, value] for key, value in change.items()]

File: procedural_solutions/numbers.c
#include <stdio.h>
#include <math.h>
#include <ctype.h>

void clear_input_buffer() {
	int c;
	while ((c = getchar()) != '\n' && c != EOF);
}

// ================== ================== ==================
double find_pi_to_nth_digit(int n), find_e_to_nth_digit(int n), tile_cost(double w, double h, double c);
void fibonacci_to_n(int n), n_fibonacci_elements(int n), prime_factors(int n), find_next_prime();
int is_prime(int n);

// ****************** ****************** ******************
int main() {
	unsigned n;
	// Find pi to nth digit
	printf("Enter the number of decimal places: ");
	if (scanf_s("%u", &n) != 1 || ferror(stdin) || n>100 ) {
		printf("Invalid input\n");
		clear_input_buffer();
		return 1;
	}
	double pi = find_pi_to_nth_digit(n);
	printf("Pi = %.*f\n", n, pi);

	// Find E to nth digit
	printf("Enter the number of decimal places: ");
	if (scanf_s("%u", &n) != 1 || ferror(stdin) || n > 100) {
		printf("Invalid input\n");
		clear_input_buffer();
		return 1;
	}
	double e = find_e_to_nth_digit(n);
	printf("E = %.*f\n", n, e);

	// Print fibonacci sequence up to n
	printf("Enter the limt: ");
	if (scanf_s("%u", &n) != 1 || ferror(stdin)) {
		printf("Invalid input\n");
		clear_input_buffer();
		return 1;
	}
	fibonacci_to_n(n);

	// Print n fibonacci sequence elements
	printf("Enter the number of elements: ");
	if (scanf_s("%u", &n) != 1 || ferror(stdin)) {
		printf("Invalid input\n");
		clear_input_buffer();
		return 1;
	}
	n_fibonacci_elements(n);

	// Find all prime factors
	printf("Enter a number greater than 2: ");
	if (scanf_s("%u", &n) != 1 || ferror(stdin) || n < 2) {
		printf("Invalid input\n");
		clear_input_buffer();
		return 1;
	}
	prime_factors(n);

	// Find prime numbers until the user chooses to stop asking for the next one
	find_next_prime();

	// Find Cost of Tile to Cover W x H Floor
	double w, h, c;
	printf("Enter width, height and cost: ");
	if (scanf_s("%lf %lf %lf", &w, &h, &c) != 3 || ferror(stdin)) {
		printf("Wrong input");
		clear_input_buffer();
		return 1;
	}
	if (w < 0 || h < 0 || c < 0) {
		printf("All numbers must be positive\n");
		return 1;
	}
	printf("Tile cost: %.2f$\n", tile_cost(w, h, c));

}

// ================== ================== ==================
double find_pi_to_nth_digit(int n) {
	double accuracy = pow(10, -n);
	double a = 1.0;
	double b = 1.0 / pow(2, 0.5);
	double t = 0.25;
	double p = 1.0;

	while (fabs(a - b) > accuracy) {
		double a_next = (a + b) / 2;
		b = pow((a * b), 0.5);
		t -= p * (pow(a - a_next, 2));
		a = a_next;
		p *= 2;
	}

	double pi = (pow(a + b, 2) / (4 * t));
	double factor = pow(10, n);
	pi = round(pi * factor) / factor;

	return pi;

}

double find_e_to_nth_digit(int n)
{
	double accuracy = pow(10, -n);
	double e = 0.0;
	double factorial = 1.0;
	double k = 0.0;

	while (1) {
		e += (1.0 / factorial);
		if ((1.0 / factorial) < accuracy) {
			break;
		}
		k += 1;
		factorial *= k;
	}
	double factor = pow(10, n);
	e = round(e * factor) / factor;
	return e;
}

void fibonacci_to_n(int n) {
	printf("Fibonacci sequence: ");

	if (n >= 0) printf("0 ");
	if (n >= 1) printf("1 ");

	int f1 = 0, f2 = 1;
	while (1) {
		int next = f1 + f2;
		if (next > n) {
			break;
		}
		printf("%d ", next);
		f1 = f2;
		f2 = next;
	}
	printf("\n");
}

void n_fibonacci_elements(int n) {
	printf("Fibonacci sequence: ");
	if (n >= 1) printf("0 ");
	if (n >= 2) printf("1 ");

	unsigned f1 = 0, f2 = 1;
	for (int i = 2; i < n; i++) {
		unsigned next = f1 + f2;
		printf("%u ", next);
		f1 = f2;
		f2 = next;
	}
	printf("\n");
}

void prime_factors(int n) {
	printf("Prime factors: ");
	while (n % 2 == 0) {
		printf("2 ");
		n /= 2;
	}
	int limit = pow(n, 0.5) + 1;
	for (int i = 3; i < limit; i += 2) {
		while (n % i == 0) {
			printf("%d ", i);
			n /= i;
		}
	}
	if (n > 2) {
		printf("%d", n);
	}
	printf("\n");
}

int is_prime(int n) {
	if (n <= 1) return 0;
	if (n == 2) return 1;
	if (n % 2 == 0) return 0;

	for (int i = 3; i * i <= n; i+=2) {
		if (n % i == 0) return 0;
	}
	return 1;
}

void find_next_prime() {
	int current = 2;
	char choice;
	printf("Press 'n' for next prime, any other key to quit\n");
	
	while (1) {
		if (is_prime(current)) {
			printf("%d is prime, continue? ", current);
			if (scanf_s("%c", &choice,1) != 1 || ferror(stdin) || (tolower(choice) != 'n')) {
				clear_input_buffer();
				break;
			}
			clear_input_buffer();
		}
		current++;
	}
	printf("\nPrime number generation stopped\n");
}

double tile_cost(double w, double h, double c) {
	return(w * h * c);
}
